src/deriveNativeScriptHash.c

```diff
@@ -0,0 +1,341 @@
+#include "deriveNativeScriptHash.h"
+#include "state.h"
+#include "uiScreens.h"
+
+enum nativeScriptType {
+	NATIVE_SCRIPT_PUBKEY = 0,
+	NATIVE_SCRIPT_ALL = 1,
+	NATIVE_SCRIPT_ANY = 2,
+	NATIVE_SCRIPT_N_OF_K = 3,
+	NATIVE_SCRIPT_INVALID_BEFORE = 4,
+	NATIVE_SCRIPT_INVALID_HEREAFTER = 5,
+};
+
+// distinguish between the types of PUBKEY
+// DEVICE_OWNED is specified by a path
+// THIRD_PARTY is specified by a pubkey hash
+enum {
+	PUBKEY_DEVICE_OWNED = 0,
+	PUBKEY_THIRD_PARTY = 1,
+};
+
+static ins_derive_native_script_hash_context_t* ctx = &(instructionState.deriveNativeScriptHashContext);
+
+// Helper functions
+
+bool expectsRequest()
```

Comment:
> All local functions should be declared `static` --- they are not visible outside of the module then.

Does this mean that also the `deriveNativeScriptHash_handle*` functions (aside from `handleAPDU`) should also be `static`?

Reply:
I guess there is no strict rule on this; afaik using static helps compiler/linker with optimization (when creating an object module for a single .c file, it is not yet clear whether the function is being called from elsewhere or not; with `static`, it is forbidden to call it from elsewhere). And you avoid the problem of implicitly declared functions (not put in header, but available to linker) which only generates a warnings which is sometimes ignored. So I don't see reasons not to make a function static if it can be.

Reply:
Okay, where it was possible I changed the function to `static`.